### ISSUE TITLE
segfault linking pandoc with text2 rc on M1 mac

### ISSUE THREAD

**Issue by jgm** [haskell/text#396]
I've successfully compiled pandoc HEAD against the text 2 release candidate, using this cabal.project on an Intel mac.
```
packages: pandoc.cabal
packages: https://hackage.haskell.org/package/text-2.0/candidate/text-2.0.tar.gz
tests: True
flags: +embed_data_files
constraints: aeson >= 2.0.1.0

allow-newer:
  attoparsec-iso8601:text,
  aws:attoparsec,
  blaze-builder:text,
  blaze-html:text,
  blaze-markup:text,
  cassava:text,
  concurrent-output:text,
  errors:text,
  feed:text,
  haddock-library:text,
  hslua-classes:text,
  hslua-core:text,
  hslua-marshalling:text,
  hslua-module-path:text,
  hslua-module-system:text,
  hslua-module-text:text,
  hslua-module-version:text,
  hslua-objectorientation:text,
  hslua-packaging:text,
  hslua:text,
  HsYAML:text,
  http-api-data:text,
  jira-wiki-markup:text,
  lens:text,
  megaparsec:text,
  pandoc-lua-marshal:text,
  pandoc:text,
  regex-tdfa:text,
  scientific:text,
  strict:text,
  tasty-lua:text,
  text-conversions:text,
  text-short:text,
  unicode-collation:text,
  uuid-types:text,
  uuid:text,
  dhall:text,
  turtle:text,
  special-values:text,
  quickcheck-instances:text,
  foldl:text,
  uri-encode:text,
  serialise:text,
  parsers:text,
  cborg-json:text

-- source-repository-package
--   type: git
--   location: https://github.com/jgm/texmath.git
--   tag: 674bcbaec03e5550f155623de6662953bd157625

source-repository-package
  type: git
  location: https://github.com/haskell/aeson

source-repository-package
  type: git
  location: https://github.com/jgm/ipynb.git
  tag: 00246af10885c2ad4413ace4f69a7e6c88297a08
```

When I try the same thing on an M1 mac, I get a segfault in the linking phase:
```
Building executable 'pandoc' for pandoc-2.16.2..
[1 of 2] Compiling Main
[2 of 2] Compiling Paths_pandoc
Linking /Users/jgm/src/pandoc/dist-newstyle/build/aarch64-osx/ghc-8.10.7/pandoc-2.16.2/x/pandoc/build/pandoc/pandoc ...
clang: error: unable to execute command: Segmentation fault: 11
clang: error: linker command failed due to signal (use -v to see invocation)
`gcc' failed in phase `Linker'. (Exit code: 254)
clang: error: unable to execute command: Segmentation fault: 11
clang: error: linker command failed due to signal (use -v to see invocation)
`gcc' failed in phase `Linker'. (Exit code: 254)
cabal: Failed to build exe:pandoc from pandoc-2.16.2.
Failed to build test:test-pandoc from pandoc-2.16.2.
```

With the current released version of text, the linking stage completes successfully on the same machine.  So it seems to be related in some way to the text2 changes.

**Comment by phadej** [comment]
@jgm, could you try with

```
packages text
  flags: -simdutf
```

just out of curiosity.

**Comment by Bodigrim** [comment]
@jgm I'm sorry, I do not have M1 hardware available.  I know that GHC build with text-2.0 succeeds on M1 hardware, so in theory it is supported. 

Could you please test with and without `simdutf` flag + with and without `bytestring-0.11.2.0`?

**Comment by Bodigrim** [comment]
@jgm + the same but with GHC 9.2.

**Comment by jgm** [comment]
I tried with
```
packages text
  flags: -simdutf
```
but also got the segfault.  I will try with that bytestring version and also with ghc 9.2, and report back.

**Comment by jgm** [comment]
With the new bytestring, I get
```
Failed to build th-lift-instances-0.1.18.
Build log (
/Users/jgm/.cabal/logs/ghc-8.10.7/th-lft-nstncs-0.1.18-81909c6d.log ):
Configuring library for th-lift-instances-0.1.18..
Preprocessing library for th-lift-instances-0.1.18..
Building library for th-lift-instances-0.1.18..
[1 of 1] Compiling Instances.TH.Lift ( src/Instances/TH/Lift.hs, dist/build/Instances/TH/Lift.o, dist/build/Instances/TH/Lift.dyn_o )

src/Instances/TH/Lift.hs:267:10: error:
    Duplicate instance declarations:
      instance Lift ByteString.ByteString
        -- Defined at src/Instances/TH/Lift.hs:267:10
      instance Lift ByteString.ByteString
        -- Defined in ‘Data.ByteString.Internal’
    |
267 | instance Lift ByteString.ByteString where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Instances/TH/Lift.hs:281:10: error:
    Duplicate instance declarations:
      instance Lift ByteString.Lazy.ByteString
        -- Defined at src/Instances/TH/Lift.hs:281:10
      instance Lift ByteString.Lazy.ByteString
        -- Defined in ‘Data.ByteString.Lazy.Internal’
    |
281 | instance Lift ByteString.Lazy.ByteString where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabal: Failed to build th-lift-instances-0.1.18 (which is required by
test:test-pandoc from pandoc-2.16.2). See the build log above for details.
```

**Comment by jgm** [comment]
Ah wait, the syntax should be
```
package text
  flags: -simdutf
```
not `packages`. I'd missed the warning about that. Let me try again.

**Comment by jgm** [comment]
With ghc 9.2 I can't seem to build `basement`.
```
Basement/Numerical/Conversion.hs:84:46: error:
    * Couldn't match expected type `Word32#' with actual type `Word#'
    * In the first argument of `W32#', namely
        `(uncheckedShiftRL# w64 32#)'
      In the first argument of `Word32x2', namely
        `(W32# (uncheckedShiftRL# w64 32#))'
      In the expression:
        Word32x2
          (W32# (uncheckedShiftRL# w64 32#)) (W32# (narrow32Word# w64))
   |
84 | word64ToWord32s (W64# w64) = Word32x2 (W32# (uncheckedShiftRL# w64 32#)) (W32# (narrow32Word# w64))
   |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^

Basement/Numerical/Conversion.hs:84:81: error:
    * Couldn't match expected type `Word32#' with actual type `Word#'
    * In the first argument of `W32#', namely `(narrow32Word# w64)'
      In the second argument of `Word32x2', namely
        `(W32# (narrow32Word# w64))'
      In the expression:
        Word32x2
          (W32# (uncheckedShiftRL# w64 32#)) (W32# (narrow32Word# w64))
   |
84 | word64ToWord32s (W64# w64) = Word32x2 (W32# (uncheckedShiftRL# w64 32#)) (W32# (narrow32Word# w64))
   |                                                                                 ^^^^^^^^^^^^^^^^^
```

**Comment by phadej** [comment]
I'll make revisions for `th-lift-instances`.

**Comment by Bodigrim** [comment]
@jgm use the following to compile with GHC 9.2

```cabal
source-repository-package
  type: git
  location: https://github.com/tfausak/hs-memory
  tag: 3cf661a8a9a8ac028df77daa88e8d65c55a3347a

source-repository-package
  type: git
  location: https://github.com/TomMD/foundation
  tag: 0bb195e1fea06d144dafc5af9a0ff79af0a5f4a0
  subdir: basement

source-repository-package
  type: git
  location: https://github.com/josephcsible/cryptonite
  tag: 3b081e3ad027b0550fc87f171dffecbb20dedafe

allow-newer:
  haddock-library:base
```

Tests are likely to fail because of known issues with GHC 9.2.1, but that's fine, because we are investigating linker error so far.

**Comment by Bodigrim** [comment]
@jgm actually, can you run `text` own test suite on M1? Does it link fine?

**Comment by jgm** [comment]
Here's what I've found:

Pandoc builds with ghc 9.2.0, given the cabal.project stanzas you suggested above.  The test suite has 118 failures (as you said to expect).

Cloning this repository and attempting `cabal test`, I get
```
Undefined symbols for architecture arm64:
  "_c6CkM_closure", referenced from:
      ltmp8337 in Text.o
      ltmp8338 in Text.o
      ltmp8339 in Text.o
      ltmp8340 in Text.o
      ltmp8341 in Text.o
      _TestsziPropertiesziText_testText46_closure in Text.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
`gcc' failed in phase `Linker'. (Exit code: 1)
```
That's with ghc 8.10.7.

With ghc 9.2.0, the build succeeds but 32 tests fail.
```
      t_utf16LE:                                  FAIL
        *** Failed! Falsified (after 13 tests and 2 shrinks):
        "\63308\&3B\151087\134938\1007848"
        Divergence: (Left Cannot decode input: Data.Text.Internal.Encoding.Fusion.streamUtf16LE: Invalid UTF-16LE stream,Right "\63308\&3B\151087\134938\1007848")
        Use --quickcheck-replay=259718 to reproduce.
        Use -p '/t_utf16LE/' to rerun this test only.
      tl_utf16LE:                                 FAIL
        *** Failed! Falsified (after 9 tests and 3 shrinks):
        "\1108868"
        "\56314\57220" /= "\1108868"
        Use --quickcheck-replay=329821 to reproduce.
        Use -p '/tl_utf16LE/' to rerun this test only.
      t_utf16BE:                                  FAIL
        *** Failed! Falsified (after 39 tests and 9 shrinks):
        "11B1\1109158\1014168\1075057\128336\1099898\809\983702\1006780"
        Divergence: (Left Cannot decode input: Data.Text.Internal.Encoding.Fusion.streamUtf16BE: Invalid UTF-16BE stream,Right "11B1\1109158\1014168\1075057\128336\1099898\809\983702\1006780")
        Use --quickcheck-replay=867350 to reproduce.
        Use -p '/t_utf16BE/' to rerun this test only.
      tl_utf16BE:                                 FAIL
        *** Failed! Falsified (after 8 tests and 3 shrinks):
        "\145016"
        "\55373\56952" /= "\145016"
        Use --quickcheck-replay=984896 to reproduce.
        Use -p '/tl_utf16BE/' to rerun this test only.
      t_unsnoc:                                   FAIL
        *** Failed! Falsified (after 4 tests and 2 shrinks):
        "\144484"
        0
        orig
        Just ("",'\144484') /= Just ("\144484",'\-2972')
        Use --quickcheck-replay=325757 to reproduce.
        Use -p '/t_unsnoc/' to rerun this test only.
      tl_unsnoc:                                  FAIL
        *** Failed! Falsified (after 6 tests and 2 shrinks):
        "\147227"
        0
        orig
        Just ("",'\147227') /= Just ("\147227",'\-229')
        Use --quickcheck-replay=72263 to reproduce.
        Use -p '/tl_unsnoc/' to rerun this test only.
      t_init:                                     FAIL
        *** Failed! Falsified (after 7 tests and 2 shrinks):
        "\71875"
        0
        orig
        "" /= "\71875"
        Use --quickcheck-replay=894299 to reproduce.
        Use -p '/basics.t_init/' to rerun this test only.
      tl_init:                                    FAIL
        *** Failed! Falsified (after 11 tests and 6 shrinks):
        "\32962"
        0
        orig
        "" /= "\32962"
        Use --quickcheck-replay=943695 to reproduce.
        Use -p '/basics.tl_init/' to rerun this test only.
      regressions
    folds-unfolds
      folds
        special
      scans
        t_scanr:                                  FAIL
          *** Failed! Falsified (after 5 tests and 8 shrinks):
          {_->'a'}
          'a'
          "\48490"
          0
          orig
          "aa" /= "aaa"
          Use --quickcheck-replay=105982 to reproduce.
          Use -p '$0=="All.Properties.folds-unfolds.scans.t_scanr"' to rerun this test only.
        t_scanr1:                                 FAIL
          *** Failed! Falsified (after 2 tests and 4 shrinks):
          {_->'a'}
          "\1011507"
          0
          orig
          "\1011507" /= "a\65532\179"
          Use --quickcheck-replay=448294 to reproduce.
          Use -p '/t_scanr1/' to rerun this test only.
        tl_scanr1:                                FAIL
          *** Failed! Falsified (after 8 tests and 6 shrinks):
          {_->'a'}
          "\1032771"
          0
          orig
          "\1032771" /= "a\131a\65481"
          Use --quickcheck-replay=925110 to reproduce.
          Use -p '/tl_scanr1/' to rerun this test only.
      mapAccum
        t_mapAccumR_char:                         FAIL
          *** Failed! Falsified (after 2 tests and 1 shrink):
          'a'
          "\1089767"
          "aa" /= "a"
          Use --quickcheck-replay=157916 to reproduce.
          Use -p '/t_mapAccumR_char/' to rerun this test only.
        t_mapAccumR:                              FAIL
          *** Failed! Falsified (after 4 tests and 7 shrinks):
          {_->(0,'a')}
          0
          "\1074450"
          0
          orig
          (0,"a") /= (0,"aa")
          Use --quickcheck-replay=304518 to reproduce.
          Use -p '$0=="All.Properties.folds-unfolds.mapAccum.t_mapAccumR"' to rerun this test only.
        tl_mapAccumR_char:                        FAIL
          *** Failed! Falsified (after 3 tests and 1 shrink):
          'a'
          "\99227"
          "aa" /= "a"
          Use --quickcheck-replay=437609 to reproduce.
          Use -p '/tl_mapAccumR_char/' to rerun this test only.
        tl_mapAccumR:                             FAIL
          *** Failed! Falsified (after 6 tests and 8 shrinks):
          {_->(0,'a')}
          0
          "\1002719"
          0
          orig
          (0,"a") /= (0,"aa")
          Use --quickcheck-replay=102429 to reproduce.
          Use -p '$0=="All.Properties.folds-unfolds.mapAccum.tl_mapAccumR"' to rerun this test only.
      unfolds
      replicate
    Text
      creation/elimination
        t_reverse_stream:                         FAIL
          *** Failed! Falsified (after 3 tests and 1 shrink):
          "\121370"
          "\3101\65512\154" /= "\121370"
          Use --quickcheck-replay=668336 to reproduce.
          Use -p '/t_reverse_stream/' to rerun this test only.
      transformations
        case conversion
        justification
      searching
      indexing
      zips
    substrings
      breaking
        t_takeEnd:                                FAIL
          *** Failed! Falsified (after 13 tests and 8 shrinks):
          1
          "\174929"
          0
          orig
          "\174929" /= "\173\145"
          Use --quickcheck-replay=464782 to reproduce.
          Use -p '/t_takeEnd/' to rerun this test only.
        t_dropEnd:                                FAIL
          *** Failed! Falsified (after 8 tests and 7 shrinks):
          1
          "\1071083"
          0
          orig
          "" /= "\1071083"
          Use --quickcheck-replay=931471 to reproduce.
          Use -p '/t_dropEnd/' to rerun this test only.
        tl_dropEnd:                               FAIL
          *** Failed! Falsified (after 7 tests and 6 shrinks):
          1
          "a\1078051"
          0
          orig
          "a" /= "a\1078051"
          Use --quickcheck-replay=923691 to reproduce.
          Use -p '$0=="All.Properties.substrings.breaking.tl_dropEnd"' to rerun this test only.
        t_takeWhileEnd:                           FAIL
          *** Failed! Exception while generating shrink-list: 'Prelude.chr: bad argument: (-3699)' (after 2 tests and 20 shrinks):
          <fun>
          "\168333"
          0
          orig
          Divergence: (Right "",Left Prelude.chr: bad argument: (-3699))
          Use --quickcheck-replay=118852 to reproduce.
          Use -p '$0=="All.Properties.substrings.breaking.t_takeWhileEnd"' to rerun this test only.
        tl_takeWhileEnd:                          FAIL
          *** Failed! Exception while generating shrink-list: 'Prelude.chr: bad argument: (-1892)' (after 3 tests and 20 shrinks):
          <fun>
          "\161948"
          0
          orig
          Divergence: (Right "",Left Prelude.chr: bad argument: (-1892))
          Use --quickcheck-replay=588275 to reproduce.
          Use -p '$0=="All.Properties.substrings.breaking.tl_takeWhileEnd"' to rerun this test only.
        t_dropWhileEnd:                           FAIL
          *** Failed! Exception while generating shrink-list: 'Prelude.chr: bad argument: (-33)' (after 2 tests and 11 shrinks):
          <fun>
          "\1023967"
          0
          orig
          Divergence: (Right "\1023967",Left Prelude.chr: bad argument: (-33))
          Use --quickcheck-replay=459110 to reproduce.
          Use -p '/t_dropWhileEnd/' to rerun this test only.
        tl_dropWhileEnd:                          FAIL
          *** Failed! Exception while generating shrink-list: 'Prelude.chr: bad argument: (-2547)' (after 11 tests and 20 shrinks):
          <fun>
          "I7\38443\19878\&8\DEL\STX\97213\99853"
          31
          orig
          Divergence: (Right "I7\38443\19878\&8\DEL\STX\97213\99853",Left Prelude.chr: bad argument: (-2547))
          Use --quickcheck-replay=960574 to reproduce.
          Use -p '/tl_dropWhileEnd/' to rerun this test only.
        t_dropAround:                             FAIL
          *** Failed! Exception while generating shrink-list: 'Prelude.chr: bad argument: (-3086)' (after 5 tests and 20 shrinks):
          <fun>
          "r\119313\9202"
          4
          orig
          Divergence: (Right "r\119313\9202",Left Prelude.chr: bad argument: (-3086))
          Use --quickcheck-replay=560816 to reproduce.
          Use -p '/t_dropAround/' to rerun this test only.
        tl_dropAround:                            FAIL
          *** Failed! Exception while generating shrink-list: 'Prelude.chr: bad argument: (-666)' (after 5 tests and 20 shrinks):
          <fun>
          "\994662"
          4
          orig
          Divergence: (Right "",Left Prelude.chr: bad argument: (-666))
          Use --quickcheck-replay=622943 to reproduce.
          Use -p '/tl_dropAround/' to rerun this test only.
      breaking many
      lines and words
      predicates
        view
    builder
      decimal
      hexadecimal
      realfloat
    lowlevel
      mul
      misc
        t_literal_surrogates:                     FAIL
          tests/Tests/Properties/LowLevel.hs:80:
          ퟿ � � � � 
          expected: "\55295 \65533 \65533 \65533 \65533 \57344"
           but got: "\65533 \65533 \65533 \65533 \65533 \57344"
          Use -p '/t_literal_surrogates/' to rerun this test only.
      input-output
        t_write_read:                             FAIL
          *** Failed! Exception: './crashy45796-7.txt: hGetContents: invalid argument (invalid byte sequence)' (after 4 tests and 2 shrinks):
          NewlineMode {inputNL = LF, outputNL = LF}
          BlockBuffering Nothing
          ["\150551"]
          Use --quickcheck-replay=540950 to reproduce.
          Use -p '$0=="All.Properties.lowlevel.input-output.t_write_read"' to rerun this test only.
        tl_write_read:                            FAIL
          *** Failed! Exception: './crashy45796-34.txt: hClose: invalid argument (Bad file descriptor)' (after 4 tests and 4 shrinks):
          NewlineMode {inputNL = LF, outputNL = LF}
          BlockBuffering Nothing
          ["\1032771"]
          Use --quickcheck-replay=206326 to reproduce.
          Use -p '$0=="All.Properties.lowlevel.input-output.tl_write_read"' to rerun this test only.
        t_write_read_line:                        FAIL
          *** Failed! Exception: './crashy45796-54.txt: hGetLine: invalid argument (invalid byte sequence)' (after 3 tests and 2 shrinks):
          NewlineMode {inputNL = CRLF, outputNL = LF}
          LineBuffering
          "\1106489"
          Use --quickcheck-replay=181280 to reproduce.
          Use -p '/t_write_read_line/' to rerun this test only.
        tl_write_read_line:                       FAIL
          *** Failed! Exception: './crashy45796-79.txt: hGetLine: invalid argument (invalid byte sequence)' (after 4 tests and 3 shrinks):
          NewlineMode {inputNL = CRLF, outputNL = CRLF}
          BlockBuffering (Just 2)
          "\140242"
          Use --quickcheck-replay=641286 to reproduce.
          Use -p '/tl_write_read_line/' to rerun this test only.
    read
  Regressions

32 out of 465 tests failed (5.67s)
```

**Comment by Bodigrim** [comment]
> With ghc 9.2.0, the build succeeds but 32 tests fail.

Yeah, that's expected, GHC 9.2 native code generator for ARM is horribly broken.

Are these results with `simdutf` flag enabled (as it is by default)?

**Comment by jgm** [comment]
> Are these results with simdutf flag enabled (as it is by default)?

No, these are with `flags: -simdutf`.  IF you want, I can try with `+simdutf`.

**Comment by Bodigrim** [comment]
I got myself an AWS virtual box. With GHC 8.10.7 `text` tests themselves fails to link:

```
$ git clone https://github.com/haskell/text
$ cabal test -w ghc-8.10
Linking /home/ubuntu/text/dist-newstyle/build/aarch64-linux/ghc-8.10.7/text-2.0/t/tests/build/tests/tests ...
/home/ubuntu/text/dist-newstyle/build/aarch64-linux/ghc-8.10.7/text-2.0/t/tests/build/tests/tests-tmp/Tests/Properties/Text.o(.text+0xefe08): error: undefined reference to 'c6CdG_closure'
/home/ubuntu/text/dist-newstyle/build/aarch64-linux/ghc-8.10.7/text-2.0/t/tests/build/tests/tests-tmp/Tests/Properties/Text.o(.text+0xefe80): error: undefined reference to 'c6CdG_closure'
/home/ubuntu/text/dist-newstyle/build/aarch64-linux/ghc-8.10.7/text-2.0/t/tests/build/tests/tests-tmp/Tests/Properties/Text.o(.text+0xefee0): error: undefined reference to 'c6CdG_closure'
/home/ubuntu/text/dist-newstyle/build/aarch64-linux/ghc-8.10.7/text-2.0/t/tests/build/tests/tests-tmp/Tests/Properties/Text.o(.text+0xeff38): error: undefined reference to 'c6CdG_closure'
collect2: error: ld returned 1 exit status
`gcc' failed in phase `Linker'. (Exit code: 1)

$ cabal test -w ghc-8.10 -f-simdutf
<same>
`gcc' failed in phase `Linker'. (Exit code: 1)
```

With GHC 9.0.1 linking (even with `simdutf` enabled) is fine and tests pass:

```
$ cabal test -w ghc-9.0
Test suite tests: PASS
```

With GHC 9.2.1 linking is fine, but tests fail because of broken GHC code generator:

```
$ cabal test -w ghc-9.2
Test suite tests: FAIL
```

That said, so far does not look like a fault on `text`'s side. I can potentially prohibit GHC 8.10 on ARM arch, so that users do not spend time trying to build it, but it might happen that an upcoming GHC 8.10.8 will fix this.

**Comment by Bodigrim** [comment]
I suspect that `undefined reference to 'c6CdG_closure'` is similar to https://gitlab.haskell.org/ghc/ghc/-/issues/20617, which is [tentatively](https://gitlab.haskell.org/ghc/homepage/-/merge_requests/96/diffs) planned for GHC 8.10.8.

**Comment by Bodigrim** [comment]
I also built `pandoc` with GHC 9.2 against `text` with `simdutf` enabled. Linking is fine, even while tests fail, as discussed above. Altogether, I'm afraid we are dealing with GHC bugs and not `text`'s ones, not much I can fix on my side unfortunately.

**Comment by jgm** [comment]
Do you understand why the new text fails to link with ghc 8.10.7? It seems unfortunate if text >= 2 won't work with ghc 8.10.7, which is the most stable  ghc version we have at the moment.

**Comment by Bodigrim** [comment]
Not really. It should be something which got fixed before GHC 9.0.1, but I do not have necessary sleuthing skills. Moreover, linking issue in `text` test suite is not necessarily the same with `pandoc` one.

`text` is a boot package, so users of GHC 8.10 always compile against `text-1.2.4.1`, unless they override it explicitly. Straightforward `cabal install pandoc` won't be affected by `text-2.0` release until GHC 9.4.

**Comment by jgm** [comment]
OK, thanks. Well, if you're satisfied that there's not a bug in text, we can close this issue.

**Comment by Bodigrim** [comment]
Using `git bisect` I learned that `text` test suite stopped to link on ARM long before UTF-8 work, in an innocently looking c8fd76fe3eb7696790fbf747f426619b1eb26d85. Apparently after enabling rewrite rules in test suite, some references became dangling. This just underlines fragility of ARM linking with GHC 8.10.7.

@jgm if you have patience to use `git bisect` to determine at which point in `text` evolution `pandoc` test suite stopped to link, I'm happy to discuss a possible workaround, but otherwise this is a GHC issue, fixed somewhere before 9.0.

**Comment by jgm** [comment]
I don't have the patience...I'll accept that it's a ghc issue.
Thanks!
